services/task/task.go

```diff
@@ -126,3 +126,25 @@ func CreateMigrateTask(doer, u *user_model.User, opts base.MigrateOptions) (*adm
 
 	return task, nil
 }
+
+// RetryMigrateTask retry a migrate task
+func RetryMigrateTask(repoID int64) error {
+	migratingTask, err := admin_model.GetMigratingTask(repoID)
+	if err != nil {
+		log.Error("GetMigratingTask: %v", err)
+		return err
+	}
+	if migratingTask.Status == structs.TaskStatusQueued || migratingTask.Status == structs.TaskStatusRunning {
+		return nil
+	}
+
+	// Reset task status and messages
+	migratingTask.Status = structs.TaskStatusQueued
+	migratingTask.Message = ""
+	if err = migratingTask.UpdateCols("status", "message"); err != nil {
+		log.Error("task.UpdateCols failed: %v", err)
+		return err
+	}
+
+	return taskQueue.Push(migratingTask)
```

Comment:
The "failed" migration might have left garbage on the storage/database, if the garbage aren't deleted, the next migration would fail again (due to the conflicts)

Reply:
Will it? I tested several timeouts in my local area and they all worked perfectly.

Reply:
IIRC the migration could have multiple steps: git clone, issue migration, etc.

What if "git clone" succeeds but "issue migration" fails?